### PR TITLE
🌱 Make release notes generation concurrent

### PR DESCRIPTION
**Pull request by g-gaston**
**What this PR does / why we need it**:
Now that we retrieve the area labels from the PRs, the release notes generation can take up to 2 minutes for big releases like 1.5. Although in general this is not a problem since it should be a one time process, it does slow down the comms release team when iterating in the notes, adding missing labels, etc. It also makes development of the script painful when testing it manually.

The default is 10 concurrent routines, but this can be changed with command flag.

The default value of 10 doesn't seem to run into rate limiting problems. I had to bump it to a 100 and run the command multiple times in a row to be able to get rate limited.

### CODE REVIEW

**Comment by g-gaston** [comment]
/hold
Builds on top of https://github.com/kubernetes-sigs/cluster-api/pull/8780, so waiting for that one to be merged first

**Comment by g-gaston** [comment]
rebased after https://github.com/kubernetes-sigs/cluster-api/pull/8780 merged
/unhold

**Review by nawazkh — APPROVED** [review]
This is an amazing update @g-gaston ! Thanks 🚀 
Just a minor nit, can we add any more comments around `results`, `jobs` and `go routines` for helping anyone reading this the first time?

File: hack/tools/release/notes.go [hunk 1]
```diff
@@ -223,53 +225,49 @@ func run() int {
...
-			body = strings.TrimPrefix(body, ":bug:")
-			body = strings.TrimPrefix(body, "🐛")
-		case strings.HasPrefix(body, ":book:"), strings.HasPrefix(body, "📖"):
-			key = documentation
-			body = strings.TrimPrefix(body, ":book:")
-			body = strings.TrimPrefix(body, "📖")
-			if strings.Contains(body, "CAEP") || strings.Contains(body, "proposal") {
-				key = proposals
+	results := make(chan processCommitResult)
+	jobs := make(chan *commit)
```
Comment (fabriziopandini) [inline hack/tools/release/notes.go:229]: nit
```suggestion
	commitCh := make(chan *commit)
```
(using another noun is a little bit confusing)

File: hack/tools/release/notes.go [hunk 2]
```diff
@@ -223,53 +225,49 @@ func run() int {
...
+	var wg sync.WaitGroup
+
+	wg.Add(*numWorkers)
+	for i := 0; i < *numWorkers; i++ {
+		go func() {
+			for commit := range jobs {
+				processed := processCommitResult{}
+				processed.prEntry, processed.err = generateNotesPREntry(commit)
+				results <- processed
```
Comment (fabriziopandini) [inline hack/tools/release/notes.go:238]:
```suggestion
				results <- generateNotesPREntry(commit)
```
if we change `generateNotesPREntry` to return a `processCommitResult`

File: hack/tools/release/notes.go [hunk 3]
```diff
@@ -346,3 +344,62 @@ func commandExists(cmd string) bool {
 	_, err := exec.LookPath(cmd)
 	return err == nil
 }
+
+type processCommitResult struct {
```
Comment (fabriziopandini) [inline hack/tools/release/notes.go:348]: what about
```suggestion
type releaseNoteEntry struct {
```
or
```suggestion
type releaseNoteItem struct {
```

File: hack/tools/release/notes.go [hunk 4]
```diff
@@ -346,3 +344,62 @@ func commandExists(cmd string) bool {
...
+	prEntry *prEntry
+	err     error
+}
+
+type prEntry struct {
```
Comment (fabriziopandini) [inline hack/tools/release/notes.go:353]: what about
```suggestion
type prInfo struct {
```

File: hack/tools/release/notes.go [hunk 5]
```diff
@@ -346,3 +344,62 @@ func commandExists(cmd string) bool {
...
+	title    string
+	section  string
+	prNumber string
+}
+
+func generateNotesPREntry(c *commit) (*prEntry, error) {
```
Comment (fabriziopandini) [inline hack/tools/release/notes.go:359]: what about
```suggestion
func generateReleaseNoteEntry(c *commit) (*prEntry, error) {
```

**Review by fabriziopandini — COMMENTED** [review]
Thanks for working at this!

File: hack/tools/release/notes.go (hunk 2 above)
Comment (g-gaston) [reply hack/tools/release/notes.go:238]: I would prefer to keep this
It makes the function agnostic from the fact that it's being called from concurrent routines. And delegates the aggregation of result an error to the routine itself.

I can add another method to do this encapsulation, but that seemed like extra overhead for saving two lines.

File: hack/tools/release/notes.go (hunk 3 above)
Comment (g-gaston) [reply hack/tools/release/notes.go:348]: renamed

File: hack/tools/release/notes.go (hunk 4 above)
Comment (g-gaston) [reply hack/tools/release/notes.go:353]: renamed

File: hack/tools/release/notes.go (hunk 5 above)
Comment (g-gaston) [reply hack/tools/release/notes.go:359]: renamed

File: hack/tools/release/notes.go (hunk 2 above)
Comment (fabriziopandini) [reply hack/tools/release/notes.go]: ack, fair

**Review by killianmuldoon — COMMENTED** [review]
Thanks for taking this on - nice speedup :slightly_smiling_face: 

/lgmt
/approve

**Comment by killianmuldoon** [comment]
@g-gaston You might want to force-push on top of this to see if that kicks easyCLA into action.

**Comment by killianmuldoon** [comment]
Thanks

/approve

**Pull request merged**